Improve documentation around block creation (#1086)
blockprotocol/blockprotocol: apps/site/src/_pages/docs/1_developing-blocks.mdx

In practice, most block developers will not need to know the lower-level details of the specifications, as the libraries we provide implement them.

## Creating a block

1.  Move to a folder where you want to create your block.
1.  Run `npx create-block-app@latest your-block-name`.
1.  Switch to your new folder: `cd [your-block-name]`.
1.  Run `yarn install && yarn dev` or `npm install && npm run dev`.
1.  Open [http://localhost:63212](http://localhost:63212/) in your browser to see the starter template.

### I want to use a different technology

> If you want to write blocks **using other technologies or frameworks**, you may pass a template parameter to the above command: `npx create-block-app@latest your-block-name --template <TEMPLATE>`. For the template, there are several options:

Reviewer: ```suggestion
If you want to write blocks **using other technologies or frameworks**, we provide three templates to choose from which allow you to define the entry point for your block in different ways. To change the template, pass a parameter to the above command as follows: `npx create-block-app@latest your-block-name --template <TEMPLATE>`. 

Currently the three templates are:
```

Author: Thanks, that's better https://github.com/blockprotocol/blockprotocol/pull/1086/commits/e482df926f68ccd35792f7ba18cdca1b1418f65f

1.  use the `react` template and use React to create your block (default).
1.  use the `html` template and import and use your additional libraries inside the `<script>` tag.

> 1.  use the `custom-element` template and use different approaches when constructing the element-i.e. use a custom element as a wrapper for anything else you like.

Reviewer 2: ```suggestion
1.  use the `custom-element` template, which is based on [Lit](https://lit.dev/). You can also use different approaches when constructing the element-i.e. use a custom element as a wrapper for anything else you like.
```

Author: [Apply suggestions from code review](https://github.com/blockprotocol/blockprotocol/pull/1086/commits/e6aac1d5c7761c14b2c63131981b8c2790641a1a)

> 1.  use the `react` template and use React to create your block (default).
> 1.  use the `html` template and import and use your additional libraries inside the `<script>` tag.
> 1.  use the `custom-element` template and use different approaches when constructing the element-i.e. use a custom element as a wrapper for anything else you like.

Reviewer: ```suggestion
1.  `react` - the default template which uses React and has a suite of helpful hooks you can use.
1.  `custom-element` - create a block defined as a custom element (also known as Web Components). You can use this with different approaches when constructing the element-i.e. use a custom element as a wrapper for anything else you like.
1.  `html` - give up some of the ecosystem to roll-your-own, import and use additional libraries inside the `<script>` tag.
```

You may also use any other language which can be transpiled to JavaScript. As an example, see [this blog post](https://hash.dev/blog/build-blocks-in-any-language) on writing a block using F#. This block uses React – you can use transpiled JavaScript in blocks defined as HTML files or custom elements too.

> The sections below assume, that you're using the `react` template.

Reviewer 2: I think the only difference which isn't called out is the folder structure – can we delete this, but instead say where we talk about the folder structure something like:

> For HTML templates, the folder structure is slightly different. The development server is in a `dev/` folder and the block entry point is `app.html`

Author: Also, the file names are different (`.js` instead of `tsx`). There is no codegen in HTML blocks and you have to insert all the things manually.

Reviewer 2: Can we call that out in the relevant place please? I guess where it talks about `block-metadata.json`

We mention the lack of TS generation already
> Once you have created the type representing the data your block needs, copy its URL, and update the `schema` property
in the `blockprotocol` object in `package.json`. In TypeScript block templates, you can then run `yarn schema` to
regenerate the types for your block.

Author: I tried to apply it in bbe1dd067d6082379d432d51c014ac09304abb8f

> The sections below assume, that you're using the `react` template.

Reviewer: ```suggestion
The sections below assume that you're using the `react` template.
```

In practice, most block developers will not need to know the lower-level details of the specifications, as the libraries we provide implement them.

## Creating a block

1.  Move to a folder where you want to create your block.
1.  Run `npx create-block-app@latest your-block-name`.
1.  Switch to your new folder: `cd [your-block-name]`.
1.  Run `yarn install && yarn dev` or `npm install && npm run dev`.
1.  Open [http://localhost:63212](http://localhost:63212/) in your browser to see the starter template.

### I want to use a different technology

> If you want to write blocks **using other technologies or frameworks**, we provide three templates to choose from which allow you to define the entry point for your block in different ways. To change the template, pass a parameter to the above command as follows: `npx create-block-app@latest your-block-name --template <TEMPLATE>`.

Reviewer 3: I do find this a bit confusing. Saying "other technologies. or frameworks" makes me want to ask... "What is the default one?!?!"

Author: Maybe we should mention `react` in `# Creating a block` @CiaranMn?

Reviewer 3: ```suggestion
By default, `create-block-app` uses React. If you want to write blocks **using another technology or framework**, we provide template options for you to choose from which allow you to define the entry point for your block in different ways. To change the template, pass a parameter to the above command as follows: `npx create-block-app@latest your-block-name --template <TEMPLATE>`.
```

Currently the three templates are:

1.  `react` - the default template which uses React and has a suite of helpful hooks you can use.
1.  `custom-element` - create a block defined as a custom element (also known as Web Components), which is based on [Lit](https://lit.dev/). You can also use different approaches when constructing the element-i.e. use a custom element as a wrapper for anything else you like.

> 1.  `html` - give up some of the ecosystem to roll-your-own, import and use additional libraries inside the `<script>` tag.

Reviewer 3: I find "give up some of the ecosystem to roll-your-own" confusing. Which ecosystem am I giving up? React (the default)? The Block Protocol's? If the latter what does that mean as I write a Block Protocol block?

Reviewer 2: ```suggestion
1.  `html` - plain HTML and JavaScript. Import and use additional libraries inside the `<script>` tag.
```

### Publish from your terminal

- Run `npm run build` or `yarn build` to create a production build of your block (it will appear in the `dist` folder)
- run `npx blockprotocol@latest publish` to generate a `.blockprotocolrc` file when prompted
- Replace the placeholder key in that file with your API key
- now run `npx blockprotocol@latest publish` again
- See your block on the Hub!

### Updating your block

You can update your published block at any time by running

> `npm run build && npx blockprotocol@latest publish` or `yarn build && npx blockprotocol@latest publish`

Reviewer: Can we swap this around, we mention `yarn` first elsewhere, and I think it's the default setup in the templates

Author: [Make ordering of npm run and yarn consistent](https://github.com/blockprotocol/blockprotocol/pull/1086/commits/a1cff771d71049eb25ec12f5479bf8874674a2af)
